### DISCUSSION TITLE
Questions about data filtering and regridding

### DISCUSSION THREAD

**Discussion by JairoVS** [pmlmodelling/nctoolkit#51] [General]
Hello! NCtoolkit, it's been a lifesaver for my analysis. I appreciate the efforts of the developers. 

I am working with satellite data of column densities (mol/cm2) of air pollutants, and I have some questions about data processing using nctoolkit. I need to set as_missing all the negatives and all values above 1e17, but at the same time, I want to fill all the empty grids using fill_na. 

1. Is it better to fill the na's first and then do the as_missing, or is it better to do it the opposite way?

2. How can I use the as_missing function to filter all the negatives? I tried using as_missing (<0), but it did not work

3. What would be a good number for the fill_na()? What happens if several contiguous grids are empty when using fill_na? 

Thank you in advance for your guidance.

**Comment by robertjwilson** [comment]
Thanks.

1. Is it better to fill the na's first and then do the as_missing, or is it better to do it the opposite way?
    **I would set the missing values first.** 

2. How can I use the as_missing function to filter all the negatives? I tried using as_missing (<0), but it did not work
    **Set a suitable range to a missing value:
    ds.as_missing([-1e90, 0])**

3. What would be a good number for the fill_na()? What happens if several contiguous grids are empty when using fill_na?
    **This is somewhat data dependent. The method is distance weighted and `n` defines how many neighbours to use. Typically, 8 covers most use cases, as you have 8 cells around each cell. `n=1` is a simple nearest-neighbour and often that is fine. It really depends on what kind of gaps you have. Sometimes I need to infill coasts, so in that case would just do `n=5`.**

**Reply by JairoVS** [reply]
Thank you very much Robert! I will do as you suggested.

**Reply by JairoVS** [reply]
One more question, when reading several files using wildcard nc.open_data("data"/*.nc4), and applying the as_missing and fill_na, is it done in all the opened files?

**Reply by robertjwilson** [reply]
Yes. It will do it on all opened files. If you want to speed it up you can run the processes in parallel by setting them number of cores to use, e.g.

    nc.options(cores = 4)

**Reply by JairoVS** [reply]
Thank you very much for your answer! One more question related to the same topic, if I have several variables in the same NetCDF and want to apply as_missing, is it possible to choose only one of the variables? Or can the method only be applied to all of the variables in the file?

**Reply by robertjwilson** [reply]
Right now you can only apply it to *all* variables. If you only want to apply it to one variable you would need to create a separate dataset including only it and then merge it back in.

**Reply by JairoVS** [reply]
Understand! I appreciate your guidance.

**Reply by JairoVS** [reply]
Hello Robert! Again, I want to ask you some questions regarding the vertical_mean function. Is it better to open a new discussion, or can I ask it here? Or is it better if I contact you through your email? Many thanks for considering my request.

**Reply by robertjwilson** [reply]
Just open a new discussion
